README.md
# Article 2 TV News

Uses python 3.4

Article2TVNews allows users to find TV news clips on subjects related
to an article they are reading.  It provides methods to extract text from an
article and get tv news clips based on the article's subject.  This package is
written for the Internet Archive's WayBack Machine and its Context browser
extension. Typical usage looks something like this::

    #!/usr/bin/env python

    from article2tvnews import ArticleExtraction, NewsRecommender

    url = <news-article-url>
    html = ArticleExtraction.getHTML(url)
    article = ArticleExtraction.extract(html)
    for clip in NewsRecommender.makeRecommendations(article):
        print(clip)


## Entity Extraction
By default, NewsRecommender uses SpaCy for entity extraction. However, one can choose to use the [OpenCalais API](http://www.opencalais.com/).  First, the user must [register to get an API key](http://www.opencalais.com/opencalais-api/).  Then pass it as `calais_token` when you call `makeRecommendations()`::

    #!/usr/bin/env python

    from article2tvnews import ArticleExtraction, NewsRecommender

    url = <news-article-url>
    html = ArticleExtraction.getHTML(url)
    article = ArticleExtraction.extract(html)
    for clip in NewsRecommender.makeRecommendations(article, calais_token = <api_key>):
        print(clip)


## Possible Changes:
- Time period for videos
- extract embedded tweets (possible added weight)


## Thanks


Thanks to the wonderful people at *The Internet Archive* for giving me this
opportunity, and a special thanks to:

* Mark

* Dvd

* Vangelis

* Anish

* And of course, Brewster Kahle


CHANGES.txt
v0.1.0, August 30, 2018 -- Initial release.
v0.1.5, October 22, 2018 -- Adds option to use SpaCy entity extraction instead of OpenCalais


test/test_NewsRecommender.py
from unittest import TestCase
from tvnews import ArticleExtraction, NewsRecommender

class TestNewsRecommender(TestCase):
    good_url = "https://www.huffingtonpost.com/entry/alex-jones-infowars-app-apple-google_us_5b694ec3e4b0de86f4a4bc1d"
    urlencode_url = "https://www.vox.com/policy-and-politics/2018/9/24/17157194/rosenstein-trump-mueller-fired-doj-fbi"

    def test_makeRecommendations_json_response(self):
        article = ArticleExtraction.extract(ArticleExtraction.getHTML(self.good_url))
        clips = NewsRecommender.makeRecommendations(article)
        for clip in clips:
            self.assertTrue(clip.get("preview_url"))
            self.assertTrue(clip.get("similarity"))
            self.assertTrue(clip.get("show"))
            self.assertTrue(clip.get("date"))
            self.assertTrue(clip.get("station"))
            self.assertTrue(clip.get("preview_thumb"))
            self.assertTrue(clip.get("snippet"))

    def test_makeRecommendations_using_urlencoding(self):
        article = ArticleExtraction.extract(ArticleExtraction.getHTML(self.urlencode_url))
        clips = NewsRecommender.makeRecommendations(article)
        for clip in clips:
            self.assertTrue(clip.get("preview_url"))
            self.assertTrue(clip.get("similarity"))
            self.assertTrue(clip.get("show"))
            self.assertTrue(clip.get("date"))
            self.assertTrue(clip.get("station"))
            self.assertTrue(clip.get("preview_thumb"))
            self.assertTrue(clip.get("snippet"))

    def test_makeRecommendations_sort(self):
        article = ArticleExtraction.extract(ArticleExtraction.getHTML(self.good_url))
        clips = NewsRecommender.makeRecommendations(article)
        for i in range(len(clips)-1):
            self.assertTrue(clips[i].get("similarity") <= clips[i+1].get("similarity"))

    def test_makeRecommendations_bad_calais(self):
        article = ArticleExtraction.extract(ArticleExtraction.getHTML(self.good_url))
        clips = NewsRecommender.makeRecommendations(article, calais_token = "klasdallksdlfaklke")
        for clip in clips:
            self.assertTrue(clip.get("preview_url"))
            self.assertTrue(clip.get("similarity"))
            self.assertTrue(clip.get("show"))
            self.assertTrue(clip.get("date"))
            self.assertTrue(clip.get("station"))
            self.assertTrue(clip.get("preview_thumb"))
            self.assertTrue(clip.get("snippet"))


test/test_ArticleExtraction.py
from unittest import TestCase
from tvnews import ArticleExtraction

class TestArticleExtraction(TestCase):
    good_url = "https://www.huffingtonpost.com/entry/alex-jones-infowars-app-apple-google_us_5b694ec3e4b0de86f4a4bc1d"
    bad_url = "https://www.huffingtonpost.com/"
    def test_getHTML_returns_string(self):
        html = ArticleExtraction.getHTML(self.good_url)
        self.assertTrue(isinstance(html, str))



    def test_extract_returns_dict(self):
        html = ArticleExtraction.getHTML(self.good_url)
        article = ArticleExtraction.extract(html)
        self.assertTrue(isinstance(article, dict))
        self.assertTrue(isinstance(article['title'], str))
        self.assertTrue(isinstance(article['body'], str))


tvnews/ArticleExtraction.py
#!/usr/bin/python3.4
# Article Extraction Module
# written by Max Reinisch

# Given an article URL, extract the article body text

import json
import urllib3
import re
from readability import Document


import logging

log = logging.getLogger(__name__)
header = {"Accept-Encoding": "gzip", "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"}
tag_re = re.compile(r'(<!--.*?-->|<[^>]*>)')
urllib3.disable_warnings()
http = urllib3.PoolManager()
logging.getLogger('readability.readability').setLevel(logging.CRITICAL)

def extract(html):
    """
    Simply uses regex and readability to get document body from html
    """

    doc = Document(html)
    json_ret = {"title": tag_re.sub('', doc.title()),
    "body": tag_re.sub('', doc.summary()).replace("\n", " " ).replace("\xa0", " ")}
    return json_ret

def getHTML(url):
    """
    Requests article and returns html
    """
    res = http.request('GET', url, headers=header)
    if res.status >=400:
        log.error("Requested URL returned status code: " +str(res.status)+ ".  Exiting...")
        exit(1)
    # log.info(res.data)
    return res.data.decode('utf-8')


tvnews/NewsRecommender.py
#!/usr/bin/python3.4
# News Recommender version 2
# written by Max Reinisch

# Given a URL for a news article, recommend TV news
# clips similar to article contents

####################
##### IMPORTS ######
####################

import urllib3
from urllib.parse import urlencode
import json
from sklearn.feature_extraction.text import TfidfVectorizer
from scipy.spatial.distance import cosine
import spacy
from collections import Counter
import en_core_web_sm
import logging



####################
#### VARIABLES #####
####################

CALAIS_ENDPOINT = 'https://api.thomsonreuters.com/permid/calais'
GDELT_HEADER = {'Content-Type' : 'application/json', 'outputformat' : 'application/json'}
log = logging.getLogger(__name__)
urllib3.disable_warnings()
http = urllib3.PoolManager()
nlp = en_core_web_sm.load()

def makeRecommendations(article, calais_token = False):
    query = getSearchQuery(article, calais_token)
    GDELT_response = getGDELTv2Response(query) #GDELT_response in JSON format
    if(GDELT_response.get('clips')):
        return list(sortClipsBySimilarity(GDELT_response.get("clips"), article))
    else:
        return []



def getSearchQuery(article, calais_token = False):
    if(calais_token):
        calais_json = getOpenCalaisResponse(article, calais_token)
        if calais_json != "SpaCy":
            return parseCalais(calais_json)
    entities = extractEntities(article)
    return parseEntities(entities)

def extractEntities(article):
    doc = nlp(article["title"] + article["title"] + article["body"])
    return [x.text for x in doc.ents]

def parseEntities(entities):
    if len(entities) == 0:
        return []

    return [e[0] for e in Counter(entities).most_common(3)]

def getOpenCalaisResponse(article, calais_token):
    CALAIS_HEADER = {'X-AG-Access-Token' : calais_token, 'Content-Type' : 'text/raw', 'outputformat' : 'application/json'}
    response = http.request('POST', CALAIS_ENDPOINT, body= (article.get("title")+" "+article.get("body")).encode('utf-8'), headers=CALAIS_HEADER, timeout=80)
    if response.status >= 400:
        log.warning("OpenCalais returned status code: " + str(response.status) + ".  Falling back to SpaCy extraction protocol...")
        return "SpaCy"
    content = response.data.decode('utf-8')

    c = json.loads(content)
    return c

def parseCalais(c):
    # This function is an unsolved problem.  How to find the best search query
    # Currently returns a list of all entities' names in order of number of mentions
    entities = [values for values in c.values() if values.get("_typeGroup") == "entities"]
    if len(entities) ==0:
        return []
    return [e.get('lastname') or e['name'] for e in sorted(entities, key=lambda x: len(x['instances']), reverse=True)]

def getGDELTv2Response(query):
    good_result = False
    entities = query[:3]
    while(not good_result):
        if len(entities) ==0:
            log.warning("No search found.  Returning empty result.")
            return {}
        query = "+".join(['"' + entity + '"' for entity in entities])
        # url = 'https://api.gdeltproject.org/api/v2/tv/tv?query='+query+ 'market:%22National%22&mode=clipgallery&format=json&datanorm=perc&timelinesmooth=0&datacomb=sep&last24=yes&timezoom=yes&TIMESPAN=14days'
        encoded_args = urlencode({'query': query+' market:"National"', 'mode':'clipgallery', 'format':'json', 'datanorm':'perc',"timelinesmooth":0, "datacomb":"sep", "last24":"yes", "timezoom":"yes", "TIMESPAN":"14days"})
        url = 'https://api.gdeltproject.org/api/v2/tv/tv?' + encoded_args
        res = http.urlopen("GET", url)


        if(res.status>=400):
            log.error("GDELT returned status code: " + str(res.status)+ ".  Exiting...")
            return {}
        try:
            gdelt_json = json.loads(res.data.decode('utf-8'))
        except ValueError:
            log.warning("Bad GDELT response, Returning empty result.");
            log.warning(url)
            gdelt_json = {}
        log.info(url)
        if len(gdelt_json.keys()) ==0:
            entities = entities[0:-1]
            log.info("GDLET returned 0 results. Simplifying search to: " + str(entities))
        else:
            good_result = True
    return gdelt_json


def sortClipsBySimilarity(clips, article):
    vectorizer = TfidfVectorizer(stop_words="english", ngram_range=(1,2) )
    bow = vectorizer.fit_transform([clip.get('snippet') for clip in clips])
    article_bow = vectorizer.transform([article.get('body')])
    cosine_distances = [cosine(vec.todense(), article_bow.todense()) for vec in bow]
    ret = []
    for clip, dist in zip(clips, cosine_distances):
        clip.update({"similarity": dist})
        ret.append(clip)
    log.info(ret)
    return sorted(ret, key = lambda x: x.get('similarity'))
